prymer/offtarget/bwa.py

```diff
@@ -284,20 +286,14 @@ def __init__(
 
         super().__init__(command=command)
 
-        # HACK ALERT
-        # This is a hack. By trial and error, pysam.AlignmentFile() will block reading unless
-        # there's at least a few records due to htslib wanting to read a few records for format
-        # auto-detection. Lame.  So a hundred queries are sent to the aligner to align enable the
-        # htslib auto-detection to complete, and for us to be able to read using pysam.
-        num_warmup: int = 100
-        for i in range(num_warmup):
-            query = Query(id=f"ignoreme:{i}", bases="A" * 100)
-            fastq_str = query.to_fastq(reverse_complement=self.reverse_complement)
-            self._subprocess.stdin.write(fastq_str)
-        self.__signal_bwa()  # forces the input to be sent to the underlying process.
-        self._reader = sam.reader(path=self._subprocess.stdout, file_type=sam.SamFileType.SAM)
-        for _ in range(num_warmup):
-            next(self._reader)
+        header = []
+        for line in self._subprocess.stdout:
+            if line.startswith("@"):
+                header.append(line)
+            if line.startswith("@PG"):
+                break
```

Comment:
Something about this worries me a bit.  It's probably not worth worrying about since it's unlikely `bwa aln` will ever change the header it outputs though.

Reply:
@tfenne I was thinking similarly.

My first local draft of this PR sent in a single sentinel FASTQ record (similar to the original "hack") with a single base to align. The queryname would be set to `ignore-me` so we would break out of this loop if `line.startswith("ignore-me")`. Would you like this instead? It would be easy to add.

The only remaining side effect of sending in dummy/sentinel records is that bwa counts how many records it aligns so if the user was able to read the stderr then it would always seem as if there were extra records being aligned.

If we keep the current approach in this PR, I can add a unit test that asserts the shape of the dictionary. If we update `bwa-aln-interactive` to change the header, then that unit test should catch the regression.

Reply:
I kept the current behavior but added an `assert` below to guard for leftover sequence header lines.

If bwa stops writing a program group, this code block will hang indefinitely (unit tests will catch that regression!)